src/anndata/compat/__init__.py

```diff
@@ -293,6 +293,26 @@ def _to_fixed_length_strings(value: np.ndarray) -> np.ndarray:
     return value.astype(new_dtype)
 
 
+Group_T = TypeVar("Group_T", bound=Union[ZarrGroup, h5py.Group])
+
+
+# TODO: This is a workaround for https://github.com/scverse/anndata/issues/874
+# See https://github.com/h5py/h5py/pull/2311#issuecomment-1734102238 for why this is done this way.
+def _require_group_write_dataframe(
+    f: Group_T, name: str, df: pd.DataFrame, *args, **kwargs
+) -> Group_T:
+    if len(df.columns) > 5_000 and isinstance(f, H5Group):
+        # actually 64kb is the limit, but this should be a conservative estimate
+        if name in f:
+            # TODO: One possible solution is to delete the group and recreate it
+            raise ValueError(
```

Comment:
`write_elem` is idempotent for the same key https://github.com/scverse/anndata/blob/0302f234fb175428fae6d831bb75ab8391c48d72/src/anndata/_io/specs/registry.py#L349-L350 so I think I will remove this.